File: 2/2.1/C++/2.1/2.1.cpp
#include <iostream>

class List {
public:
    virtual bool isEmpty() = 0;
    virtual void insert(int value, int position) = 0;
    virtual int locate(int value) = 0;
    virtual int retrieve(int position) = 0;
    virtual void print() = 0;
    virtual void del(int position) = 0;
    virtual int end() = 0;
    virtual int getCurrentPosition() = 0;
    virtual ~List() {};
};

struct Cell
{
    int value;
    Cell *previous, *next;
};

class PointerList : public List
{
    Cell *first, *current, *last;
    int currentPosition;
    int elementCount;

public:
    PointerList() {
        elementCount = 0;
        currentPosition = -1;
        first = current = last = nullptr;
    }
    ~PointerList() {
        current = first;
        while (current != nullptr)
        {
            Cell* tmp = current;
            current = current->next;
            delete tmp;
        }
    }
    bool isEmpty() {
        return elementCount == 0;
    }
    int end()
    {
        return elementCount;
    }

    void insert(int value, int position) {
        if (isEmpty()) {
            elementCount++;
            first = last = current = new Cell();
            currentPosition = 0;
            current->value = value;
            current->next = nullptr;
            current->previous = nullptr;
        }
        else if (position >= end())
        {
            current = last;
            currentPosition = end();
            last = new Cell();
            last->value = value;
            last->next = nullptr;
            last->previous = current;
            current->next = last;
            elementCount++;
            current = last;
        }
        else
        {
            Cell* newCell = new Cell();
            newCell->value = value;
            newCell->previous = current;
            newCell->next = current->next;
            current->next->previous = newCell;
            current->next = newCell;
            current = newCell;
            currentPosition++;
            elementCount++;
        }

    }
    int locate(int value) {
        current = first;
        currentPosition = 0;
        while (current != nullptr)
        {
            if (current->value == value)
                return currentPosition;
            current = current->next;
            currentPosition++;
        }
        return end();

    }
    int retrieve(int position) {
        current = first;
        currentPosition = 0;
        while (current != nullptr)
        {
            if (currentPosition == position)
                return current->value;
            current = current->next;
            currentPosition++;
        }
        return -1;
    }
    void del(int position) {
        if (position == 0)
        {
            current = first->next;
            delete first;
            first = current;
        }
        else
        {
            current = first;
            currentPosition = 0;
            while (current != nullptr)
            {
                if (currentPosition == position)
                {
                    if (current->next == nullptr)
                    {
                        current = current->previous;
                        currentPosition--;
                        elementCount--;
                        delete current->next;
                        current->next = nullptr;
                        return;
                    }
                    else
                    {
                        Cell *toDelete = current;
                        current->next->previous = current->previous;
                        current->previous->next = current->next;
                        elementCount--;
                        delete toDelete;
                        return;
                    }               
                }                  
                current = current->next;
                currentPosition++;
            }
        }      
    }
    
    int getCurrentPosition()
    {
        return currentPosition;
    }
    
    void print(){
        if (!isEmpty())
        {
            for (int i = 0; i < end(); i++)
                std::cout << retrieve(i);
        }
        std::cout << std::endl;
    }

};

void purge(List *L)
{
    int p, q;
    p = 0;
    while (p != L->end())
    {
        q = p + 1;
        while (q != L->end())
        {
            if (L->retrieve(p) == L->retrieve(q))
                L->del(q);
            else
                q++;
        }
        p++;
    }
}

int main()
{
    {
        List* L = new PointerList();
        L->insert(1, 0);
        L->insert(2, L->end());
        L->insert(2, L->end());
        L->insert(3, L->end());

        L->print();
        purge(L);

        L->print();
    }

    {
        List* L = new PointerList();
        L->insert(1, 0);
        L->insert(1, L->end());
        L->insert(2, L->end());
        L->insert(3, L->end());

        L->print();
        purge(L);

        L->print();
    }

    {
        List* L = new PointerList();
        L->insert(1, 0);
        L->insert(2, L->end());
        L->insert(3, L->end());
        L->insert(3, L->end());

        L->print();
        purge(L);

        L->print();
    }
}

File: 2/2.1/Python/2.1.py
class List:
    def isEmpty(self)->bool:
        pass
    def insert(self, value: int, position: int):
        pass
    def locate(self, value: int) -> int:
        pass
    def retrieve(self, position: int) -> int:
        pass
    def print(self):
        pass
    def delete(self, position: int):
        pass
    def end(self)->int:
        pass
    def getCurrentPosition(self)->int:
        pass
class Cell:
    def __init__(self):
        self.value = None
        self.previous = None
        self.next = None
class PointerList(List):
    def __init__(self):
        self.first = None
        self.current = None
        self.last = None
        self.currentPosition = -1
        self.elementCount = 0
    def isEmpty(self) -> bool:
        return self.elementCount == 0
    def end(self) -> int:
         return self.elementCount
    def insert(self, value: int, position: int):
        if self.isEmpty():
            self.elementCount = self.elementCount + 1
            self.first = self.last = self.current = Cell()
            self.currentPosition = 0
            self.current.value = value
            self.current.next = None
            self.current.previous = None
        elif position >= self.end():
            self.current = self.last
            self.currentPosition = self.end()
            self.last = Cell()
            self.last.value = value
            self.last.next = None
            self.last.previous = self.current
            self.current.next = self.last
            self.elementCount = self.elementCount + 1
            self.current = self.last
        else:
            newCell = Cell()
            newCell.value = value
            newCell.previous = self.current
            newCell.next = self.current.next
            self.current.next.previous = newCell
            self.current.next = newCell
            self.current = newCell
            self.currentPosition = self.currentPosition + 1
            self.elementCount = self.elementCount + 1
    def locate(self, value: int) -> int:
        self.current = self.first
        self.currentPosition = 0
        while (self.current != None):
            if (self.current.value == value):
                return self.currentPosition
            self.current = self.current.next
            self.currentPosition = self.currentPosition + 1
        return self.end()
    def retrieve(self, position: int) -> int:
        self.current = self.first
        self.currentPosition = 0
        while (self.current != None):
            if (self.currentPosition == position):
                return self.current.value
            self.current = self.current.next
            self.currentPosition = self.currentPosition + 1
        return -1
    def delete(self, position: int):
        if position == 0:
            self.current = self.first.next
            self.first = self.current
        else:
            self.current = self.first
            self.currentPosition = 0
            while self.current != None:
                if self.currentPosition == position:
                    if self.current.next == None:
                        self.current = self.current.previous
                        self.currentPosition = self.currentPosition - 1
                        self.elementCount = self.elementCount - 1
                        self.current.next = None
                        return
                    else:
                        self.current.next.previous = self.current.previous
                        self.current.previous.next = self.current.next
                        self.elementCount = self.elementCount - 1
                        return

                self.current = self.current.next
                self.currentPosition = self.currentPosition + 1
    def getCurrentPosition(self)->int:
        return self.currentPosition
    def printList(self):
        if not self.isEmpty():
            for i in range(0, self.end()):
                print(self.retrieve(i))
        print("\n")


def purge(L: List):
    p = 0
    while p != L.end():
        q = p + 1
        while q != L.end():
            if L.retrieve(p) == L.retrieve(q):
                L.delete (q)
            else:
                q = q + 1
        p = p + 1


if __name__ == '__main__':

        L = PointerList()
        L.insert(1, 0)
        L.insert(2, L.end())
        L.insert(2, L.end())
        L.insert(3, L.end())

        L.printList()
        purge(L)

        L.printList()

        L = PointerList()
        L.insert(1, 0)
        L.insert(1, L.end())
        L.insert(2, L.end())
        L.insert(3, L.end())

        L.printList()
        purge(L)

        L.printList()

        L = PointerList()
        L.insert(1, 0)
        L.insert(2, L.end())
        L.insert(3, L.end())
        L.insert(3, L.end())

        L.printList()
        purge(L)

        L.printList()
